### PR TITLE
Refactor build package

### PR DESCRIPTION
**Pull request by errordeveloper**
#1430 moved code was moved from `build/build.go`, however it is still a very long file with deeply nested logic. This change attemts to make a few improvements.

### CODE REVIEW

File: builder/node.go [hunk 1]
```diff
@@ -34,9 +34,27 @@ func (b *Builder) Nodes() []Node {
...
+	if len(out) > 0 {
+		return out, nil
+	}
+	return nil, err
+}
+
 // LoadNodes loads and returns nodes for this builder.
 // TODO: this should be a method on a Node object and lazy load data for each driver.
-func (b *Builder) LoadNodes(ctx context.Context, withData bool) (_ []Node, err error) {
+func (b *Builder) LoadNodes(ctx context.Context, withData, onlyAvailable bool) (_ []Node, err error) {
```
Comment (jedevc) [inline builder/node.go:57]: Ideally, I'd like to not modify this to take another boolean.

`AvailableNodes` can be renamed to `LoadAvailableNodes` - then `LoadNodes` and `LoadAvailableNodes` can call `LoadNodes` internally, and do the filtering on that directly.

Comment (errordeveloper) [reply builder/node.go:57]: I think that makese sense, thanks 👍

Comment (errordeveloper) [reply builder/node.go]: PTAL [32c9bf0](https://github.com/docker/buildx/pull/1466/commits/32c9bf0683cf25f2d472c2fe939dce7daf9d774e)

File: build/build.go [hunk 2]
```diff
@@ -356,7 +81,7 @@ func toRepoOnly(in string) (string, error) {
 	return strings.Join(out, ","), nil
 }
 
-func toSolveOpt(ctx context.Context, node builder.Node, multiDriver bool, opt Options, bopts gateway.BuildOpts, configDir string, pw progress.Writer, dl dockerLoadCallback) (solveOpt *client.SolveOpt, release func(), err error) {
+func toSolveOpt(ctx context.Context, node builder.Node, multiDriver bool, opt options.Options, bopts gateway.BuildOpts, configDir string, pw progress.Writer, dl dockerLoadCallback) (solveOpt *client.SolveOpt, release func(), err error) {
```
Comment (errordeveloper) [inline build/build.go:84]: I'd like to move this function into `options` package, but it looks like that would be more invloved, so I'll leave it for another PR.

File: options/options.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
+package options
```
Comment (tonistiigi) [inline options/options.go:1]: This should not be in a separate package. If there is some import case where only types are needed then it should be types pkg under `build`. But I don't see how these cases should be possible as these are just the types that `build` pkg users for its functions and not for anything else.

**Review by tonistiigi — COMMENTED** [review]
Not sure how much this collides with #1296, but if it does we should get that in before moving a lot of things around.

File: options/options.go (hunk 3 above)
Comment (errordeveloper) [reply options/options.go:1]: > Not sure how much this collides with #1296, but if it does we should get that in before moving a lot of things around.

That PR only touches 3 lines in `build/build.go`.

**Comment by errordeveloper** [comment]
> Not sure how much this collides with #1296, but if it does we should get that in before moving a lot of things around.

That PR only touches 3 lines in `build/build.go`.

File: options/options.go (hunk 3 above)
Comment (errordeveloper) [reply options/options.go:1]: Actually, the intention was to also move `toSolveOpts` and decouple it from the driver abstraction, which I'm planning to embark on next. The intention would be to make a cleaner abstractions we can continue to iterrate on, right now `build` package has lots of things leaky pieces. Eventully such `options` package could provide a more high-level `SovleOpts` replacement.

Comment (tonistiigi) [reply options/options.go:1]: In addition to https://github.com/docker/buildx/issues/1483#issuecomment-1362011227

I'm quite skeptical of the `toSolveOpt` abstraction that wouldn't just be a gimmick that only sets some of the properties, leaving the complicated ones to `build` pkg internals. There are lots of places where build invocations directly interact with driver interface (booting, caps detection, artifact loading, result chaining, multi-node result joining etc.). What we don't want is that there is an extra abstraction that Buildx itself doesn't really use and now when we need to implement some important user-facing features like https://github.com/docker/buildx/issues/1377 or https://github.com/docker/buildx/issues/1104 (that I think both change quite a lot how `build` pkg is called), we find out that this extra abstraction does not work in this cases and does not let us implement these features.

Assuming the abstraction that doesn't have the described maintenance issues exists and it is worth investing in writing one, it should happen within the `build` pkg. The problem isn't with splitting bigger files into smaller ones or changing function signatures where it makes sense, but introducing new pkg boundaries where the pkg is not usable on its own. If after the `build` pkg refactoring, we end up in the state where public functions(files) can be split in a way they become useful independently and without sacrificing future dev cost, then it is cheap to do the pkg split.

**Comment by crazy-max** [comment]
With the controller implementation I think we can close this one?

**Pull request closed without merging**